### ISSUE TITLE
memory allocation of 41/47 bytes failed

### ISSUE THREAD

**Issue by dureuill** [meilisearch/meilisearch#4692]
**Describe the bug**

Meilisearch fails with:

```
May 27 05:48:14 ip-172-31-32-128 meilisearch[721]: memory allocation of 47 bytes failed
May 27 05:48:14 ip-172-31-32-128 systemd[1]: meilisearch.service: Main process exited, code=killed, status=6/ABRT
May 27 05:48:14 ip-172-31-32-128 systemd[1]: meilisearch.service: Failed with result 'signal'.
```

or

```
2024-06-03T04:11:44.232993Z ERROR meilisearch: info=panicked at library/std/src/sys/pal/unix/stack_overflow.rs:168:13: failed to set up alternative stack guard page: Cannot allocate memory (os error 12)
2024-06-03T04:11:44.233652Z ERROR meilisearch: info=panicked at library/core/src/panicking.rs:156:5: panic in a function that cannot unwind
```

More information:

- https://discord.com/channels/1006923006964154428/1244531390071177216/1244531390071177216
- https://discord.com/channels/1006923006964154428/1247135290049429576/1247135290049429576

**To Reproduce**

- Unknown at this point. Affected users seem to have >18 indexes and up to 2000

**Expected behavior**

- No Meilisearch crash

**Screenshots**
- see https://discord.com/channels/1006923006964154428/1247135290049429576/1247173874530127938
- https://discord.com/channels/1006923006964154428/1244531390071177216/1245354847104925706

**Meilisearch version:**
v1.8.0 and v1.8.1.
It is unclear at this point if v1.7.x are affected as well.

**Additional context**
- Happens on AWS or in an Ubuntu VM
- Hypotheses that were tested already and are unlikely to be the root cause of the bug:
  - virtual memory management for multiple indexes
  - mimalloc upgrade

**Comment by dureuill** [comment]
Next steps:

1. try and reproduce the issue to further debug with debug builds and hooks

**Comment by dureuill** [comment]
https://github.com/orgs/meilisearch/discussions/677#discussioncomment-9562242 could also be related

**Comment by dureuill** [comment]
Some progress in the investigation, thanks to the work of users in the Discord.

- it looks like `vm.map_count` increases without decrease
- the increase in maps is also accompanied by a memory increase that looks like a memory leak
- the error message alternate with an unrecoverable panic while setting the stack guard page

@irevoire  has been trying to reproduce the issue on our end, more information about whether that attempt was successful to follow today.

**Comment by iamriajul** [comment]
> https://github.com/orgs/meilisearch/discussions/677#discussioncomment-9562242 could also be related

Thanks, I found this. Recently after a server migration, I migrated meilisearch to 1.8.1 from 1.3.5 and then nightmare started, I couldn't find the reason, then suspected docker resources limit issue (https://github.com/meilisearch/meilisearch/issues/4686) but after few days today meilisearch crashed again!!! So this saved me, by just downgrading to 1.3.5.

**Comment by dureuill** [comment]
Hello @iamriajul, you can also downgrade to v1.7.6 which doesn't exhibit this behavior (v1.8 regression) but still has much better performance and a lot more features than v1.3.5

We're trying to understand the cause of the regression in v1.8. It looks like a memory leak indeed.

**Comment by dureuill** [comment]
Some news on this issue:

- @irevoire managed a reproduction of the leak on Meilisearch v1.8.0 and checked there was no reproduction in v1.7.6

<details>
<summary>v1.8.0, number of memory map according to time</summary>


<img width="653" alt="v1 8 0" src="https://github.com/meilisearch/meilisearch/assets/41078892/1a36b42e-6f91-4339-b8f6-d1bc0beff2ab">



- The number of maps is linear in time.
- At the time the machine was stopped it had reached around 16k maps

</details>



<details>
<summary>v1.7.6, number of memory maps according to time</summary>

<img width="651" alt="v1 7 6" src="https://github.com/meilisearch/meilisearch/assets/41078892/46211422-0675-4d45-a854-065f3e924381"></img>

- The number of maps is constant in time
- it hovers around 400 maps

</details>

- @irevoire tested with a downgrade of charabia from v0.8.10 to v0.8.7 (the version it was in v1.7.6). It seems to fix the leak. We're now investigating the source of the memory leak in charabia


<details>
<summary>v1.8.0 with downgraded charabia, number of memory maps according to time</summary>

![v1 8 0 with downgraded charabia](https://github.com/meilisearch/meilisearch/assets/41078892/935ea974-d65b-4c29-9848-65bbd03d95e0)


- The number of maps is constant in time
- it hovers around 400 maps

</details>

**Comment by dureuill** [comment]
⚠️ There was a setup issue in the previous test campaign; the leak is still present with charabia v0.8.7 and so was not introduced by charabia v0.8.10.

Continuing investigation...

**Comment by dureuill** [comment]
Hello,

we found a fix, see #4707 for details

**Comment by dureuill** [comment]
Fixed in v1.8.3
